src/Conversion/KrnlToAffine/KrnlToAffine.cpp

```diff
@@ -25,6 +25,215 @@
 using namespace mlir;
 
 namespace {
+// Since Krnl Dialect allows optimizations to be specified in the form of
+// recipes without being applied, some IR block may exist under Krnl loops
+// corresponding to loops that will be materialized only after relevant
+// optimization recipes are applied; these Krnl loops serve as anchors for IR
+// placement as we progressively apply optimization recipes, creating new
+// concrete loops that will correspond to these optimized loop references.
+// Whenever a concrete loop gets materialized and is referred to by Krnl loop
+// reference %loop_ref, we will need to maintain the relative positioning of IR
+// block and their parent loop operations; we do so by moving IR blocks while
+// Krnl Dialect lowering proceeds.
+//
+// Consider the following example, where we specify the recepie for a
+// 2-dimensional tiled loop, and insert memory allocation/deallocation aimed to
+// set up and clean up per-tile temporary buffer:
+//
+// %ii, %ij = krnl.define_loops 2
+// %ib, %il = krnl.block %ii 5 : (!krnl.loop) -> (!krnl.loop, !krnl.loop)
+// %jb, %jl = krnl.block %ij 4 : (!krnl.loop) -> (!krnl.loop, !krnl.loop)
+// krnl.permute(%ib, %il, %jb, %jl) [0, 2, 1, 3] : !krnl.loop, !krnl.loop,
+//     !krnl.loop, !krnl.loop
+// krnl.iterate(%ib, %jb) with (%ii -> %i = 0 to 10, %ij -> %j = 0 to 20) {
+//   %alloc = alloc() : memref<10 x f32>
+//   krnl.iterate(%il, %jl) with () {
+//     %foo = addi %i, %j : index
+//   }
+//   dealloc %alloc : memref<10 x f32>
+//  }
+//
```

Comment:
Another question: with krnl.matmul (which compute a panel of NxMxK), it will implement something akin to the 
```mlir
   krnl.iterate(%il, %jl) with () {
      %foo = addi %i, %j : index
    }
```
code above. So is it ok not to omit this inner iterate and generate the krnl.matmul in its place? Again, what does it say about the permute?

Reply:
> So is it ok not to omit this inner iterate and generate the krnl.matmul in its place?

You mean is it okay to **omit** this inner iterate right? The best way to implement this is to lay this `krnl.matmul` down as an optimization recipe, and let the krnl dialect lowering process to do the substitution for you, as opposed to modifying the code directly (which amounts to doing part of the lowering outside Krnl lowering).

This means there's probably a need to implement another optimization primitive called `substitute_with_kernel`.

Reply:
Just to update for record-keeping, per our latest discussion, directly substituting the inner loop with kernel makes more intuitive sense.